Initial go at localizing some components

chandlerprall opened the pull request:
Converts a number of text locations to use `EuiI18n`.

Couple notes:
* we need a lint rule to validate usage outside of TypeScript
* As @pugnascotia expected, I did hit a couple locations where I'd expect the localization will affect the placement of variables. Specifically these are (& linked to the change)
  * [EuiSuperSelect](https://github.com/chandlerprall/eui/blob/feature/i18n-tokens/src/components/form/super_select/super_select.js#L249)
  * [EuiComboBoxPill](https://github.com/chandlerprall/eui/blob/feature/i18n-tokens/src/components/combo_box/combo_box_input/combo_box_pill.js#L51)
  * [EuiComboBoxOptionsList](https://github.com/chandlerprall/eui/blob/feature/i18n-tokens/src/components/combo_box/combo_box_options_list/combo_box_options_list.js#L154)
  
I updated `EuiI18n`'s _default_ prop to accept a function which consumes a passed _values_ object and returns the formatted text/element. The other option (as Rory suggested) is have a formatter function be exposed through the `EuiContext`. This would let the default messages be forwarded directly to e.g. `react-intl`. I tried bringing `react-intl` directly into EUI to allow defaults to be processed by it for a cleaner API, but it doesn't look like you can get JSX in messages without directly using their `FormattedMessage` component. The other way to go instead is to throw away `EuiI18n` and use `react-intl` directly in EUI, with the hard dependency it brings - the API surface is largely the same, basically replacing _EuiI18n_ -> _FormattedMessage_.

Thoughts @snide @pugnascotia ?

snide commented:
@chandlerprall and I chatted briefly over Slack on this. He's gonna reach out to people on the downstream consumer teams to get feedback (since I'm the least helpful here). There's no need to rush into decisions and we can wait a day before going nuts in the repo.

Exciting that even given the syntax questions, we're able to cover so much of the repo so quickly. That's always a good sign :)

pugnascotia commented:
So where are we at with `react-intl`? It's attractive to have less code to maintain, especially when the APIs seem to be converging.

chandlerprall commented:
EUI will not be including `react-intl` - there is a necessity to keep our implementation untangled from whatever solution Kibana may be using. I've updated this PR to solidify the described methods of formatting localized strings, and have also added (but need to update the PR description) support for placeholders in those strings, e.g.

```javascript
<EuiI18n
  token=""
  default="{greeting}, {name}"
  values={{ greeting: 'Hello', name: 'Rory' }}
/>
```

This PR is technically ready for review but blocked for merging by https://github.com/elastic/kibana/pull/24707 to add enzyme support for React's new context API.

chandlerprall commented:
@pugnascotia this is fully ready for review. The conflicts are around the original I18n revert which is now ready to be re-reverted.

pugnascotia commented on src/components/code_editor/code_editor.js:
```diff
@@ -159,11 +160,19 @@ export class EuiCodeEditor extends Component {
-          When you&rsquo;re done, press Escape to stop {activity}.
+          <EuiI18n
+            token={`euiCodeEditor.stop${activityLabel}`}
+            default={` When you're done, press Escape to stop ${activity}.`}
```
In case whitespace matters:

```suggestion
            default={`When you're done, press Escape to stop ${activity}.`}
```

pugnascotia commented on src/components/combo_box/combo_box_options_list/combo_box_options_list.js:
```diff
@@ -139,22 +140,48 @@ export class EuiComboBoxOptionsList extends Component {
+              <EuiI18n
+                token="euiComboBoxOptionsList.createCustomOption"
+                default="Hit {key} to add {searchValue} as a custom option"
+                values={{ key: <EuiCode>ENTER</EuiCode>, searchValue: <strong>{searchValue}</strong> }}
```
Should `ENTER` be localised?

chandlerprall replied:
It is still localizable, the consumer can ignore that default value in their formatting:

```
'euiComboBoxOptionsList.createCustomOption': ({ searchValue }) => {
  return i18n.translate(
    'common.ui.euiComboBoxOptionsList.createCustomOption',
    {
      defaultMessage: 'Hit <EuiCode>ENTER</EuiCode> to add {searchValue} as a custom option',
      values: { searchValue // forwarded from EUI's call into this function }
    }
  );
}
```

Since react-intl allows JSX in the translations, the key value can be ignored and the localized value put directly into the message.

pugnascotia commented on src/components/form/super_select/super_select.js:
```diff
@@ -245,8 +246,12 @@ export class EuiSuperSelect extends Component {
-            Use the up and down keys to navigate or escape to close.
+            <EuiI18n
+              token="euiSuperSelect.screenReaderAnnouncement"
+              default={`You are in a form selector of {optionsCount} items and must select a single option.
```
Ooof, not handling item/items? :-P

chandlerprall replied:
Wasn't going to refactor any logic as part of this :-D

chandlerprall commented on src/components/code_editor/code_editor.js:
```diff
@@ -141,11 +142,11 @@ export class EuiCodeEditor extends Component {
+    const activityLabel = isReadOnly ? 'Interacting' : 'Editing';
     const activity = isReadOnly
       ? 'interacting with the code'
       : 'editing';
```
No, `activityLabel` is used in forming the tokens later. However, how I'm using it here breaks the planned static analysis of tokens, I'll refactor.

pugnascotia commented on src/components/i18n/i18n_util.tsx:
```diff
@@ -0,0 +1,99 @@
+      appendValueToChildren(child);
+      child = {propName: ''};
+    } else if (char === '}') {
+      // @ts-ignore
```
How come?

chandlerprall replied:
Updated code to use a type cast instead of completely ignoring the line to better communicate intent.

chandlerprall commented on src/components/i18n/i18n.tsx:
```diff
@@ -1,35 +1,64 @@
+  values?: I18nTokenShape<T>['values']
+): ReactChild {
+  const renderable = (i18nMapping && i18nMapping[token]) || valueDefault;
+  if (typeof renderable === 'function') {
```
Updated to remove `else` blocks.
